Feat/#2 - 데이터 파이프라인 구현

jjungyeun opened the pull request:
### 변경사항
**AS-IS**
- 데이터 수집/전처리/저장 로직 구현
- 특정 profile로 서버 실행 시 MySQL에 데이터 저장하는 클래스 추가 (테스트용 데이터)
- Bulk Insert를 위해 `JdbcTemplate`으로 native query 실행하는 로직 구현
- 기존에 존재하던 맛집 정보 업데이트 로직 구현 (폐업 시 삭제)
- 스케줄러 등록 (매일 오전 2시 실행)
- 간단한 테스트 코드 작성

**TO-BE**
- Unit Test 작성


### 테스트
- [ ] 테스트 코드

jjungyeun committed:
chore #2 - WebClient 사용을 위한 WebFlux 의존성 추가

jjungyeun committed:
feat #2 - OpenApi 설정 및 요청용 인터페이스 추가

- 사용할 open api의 기본 주소 및 서비스키 설정 추가
- api 요청용 인터페이스 정의 및 config로 빈 등록

jjungyeun committed:
feat #2 - Raw 데이터 파싱용 컨버터 구현

- raw 데이터를 담을 클래스 정의
- 맛집 분류를 구분하기 위한 Enum 클래스 정의 (파싱 시 필요)
- 응답 상태를 담고있는 head 부분과 맛집 정보를 담고 있는 row 부분 파싱 로직 구현

jjungyeun committed:
feat #2 - 데이터 수집 및 전처리 로직 구현

- API를 호출해서 데이터를 수집하는 메소드 구현
- raw 데이터 필터링 및 전처리 메소드 구현
- raw 데이터를 엔티티로 변환하는 메소드 구현

jjungyeun committed:
feat #2 - 데이터 파이프라인 로직 구현

- 데이터 수집 및 전처리 메소드를 활용하여 업종별 데이터를 수집해서 엔티티로 변환
- 변환된 엔티티의 중복을 검증하고 저장하는 로직 구현

jjungyeun committed:
feat #2 - 맛집 데이터 Bulk Insert 구현

- ID 생성 전략이 IDENTITY인 경우 saveAll() 사용 시 bulk 처리 되지 않는 문제
- JdbcTemplate을 사용해 직접 쿼리를 작성해서 bulk insert 처리

jjungyeun committed:
test #2 - 테스트 데이터 저장용 클래스 추가

- Active-profile을 "test-data"로 실행하면 연결된 MySQL에 데이터가 저장됨

jjungyeun committed:
fix #2 - bulk insert 시 중복 row 무시하도록 설정

- INSERT IGNORE 설정을 통해 bulk 내에 중복되는 데이터가 있으면 무시함

jjungyeun committed:
feat #2 - WebClient 예외 처리 로직 추가

- WebClient로 호출한 결과가 4xx 또는 5xx인 경우 예외를 던지도록 설정 추가
- 예외 처리 로직 추가 (최상단까지 예외를 넘기지 않고, 에러 로그만 남기고 계속 데이터 수집 진행)

jjungyeun committed:
Merge branch 'develop' into feat/#2

# Conflicts:
#	src/main/java/com/wanted/teamV/repository/RestaurantRepository.java

jjungyeun committed:
feat #2 - 맛집 데이터 저장 성능 개선

- 중복 데이터를 확인하던 로직이 가장 오래걸리는 부분이었음
- 어차피 insert 시에 ignore문을 걸어놨기 때문에, 애플리케이션 단에서 굳이 중복을 확인할 필요 없음
- findAllByNameAndRoadnameAddress()로 중복 체크하는 로직 제거
- 단, 반드시 DB 테이블에 UNIQUE INDEX를 걸어야 함

jjungyeun committed:
feat #2 - 데이터 파이프라인 로직에서 Transactional 어노테이션 제거

- 중간에 오류가 발생하더라도 그 전까지 저장한 데이터는 살려두는 것이 효율적일 것으로 판단하여 @Transactional 제거

jjungyeun committed:
feat #2 - OPEN API 예외 처리 로직 추가

- OPEN API에서 제공하는 예외 코드 및 메세지를 파싱하는 로직 추가

jjungyeun committed:
Merge branch 'develop' into feat/#2

# Conflicts:
#	build.gradle
#	src/main/resources/application.yml

jjungyeun committed:
test #2 - 데이터 파이프라인 테스트

- 테스트 설정 파일에 open api 관련 변수 추가
- h2에서도 돌아가도록 native query 모두 소문자로 수정
- open api에서 불러온 데이터를 저장하는 테스트 코드 추가

jjungyeun committed:
chore #2 - OPEN API 서비스키 할당하는 job 추가

jjungyeun committed:
chore #2 - OPEN API 서비스키 할당하는 job 추가

- application.yml 대신 application-test.yml에 할당하도록 수정

nayoonk928 commented:
코드 잘 봤습니다!
데이터 처리시에 만약 이미 있는 맛집인데 영업상태가 "폐업"으로 변경되었다면 db에서 제거되는 로직이 필요할 것 같아요.

jjungyeun committed:
feat #2 - 데이터 파이프라인 스케줄링 등록

- 스케줄 어노테이션 사용을 위해 @EnableScheduling 추가
- 데이터 파이프라인 메소드에 @Schedlued 추가 및 매일 오전 2시에 실행되도록 설정

jjungyeun committed:
feat #2 - 맛집 정보 업데이트 로직 구현

- 기존에 존재하던 맛집의 정보를 다시 받아온 경우 정보 업데이트
- 기존에 존재하던 맛집이 폐업한 경우 DB에서 평가와 함께 삭제
- Restaurant 엔티티에 update 메소드 추가
- 특정 맛집의 모든 평가를 지우는 쿼리 메소드 정의

jjungyeun committed:
chore #2 - 테스트 시 SQL 로그 제외

- 데이터 파이프라인 수행 시 SQL 로그가 매우 많아 Github Actions Report에서 오류 발생함

nswon reviewed (approved):
👍 나중에 다시한번 자세히 보겠습니다!

jjungyeun committed:
chore #2 - 스케줄링 시간을 설정파일로 이동

- application.yml에 스케줄링 시간 설정 추가

jjungyeun committed:
test #2 - 스케줄링 테스트

- 스케줄링 테스트를 위한 Awaitility 라이브러리 의존성 추가
- test 설정 파일에 스케줄링 시간 설정 추가
- 10초마다 스케줄링이 시작하도록 설정 후 3분 내에 데이터가 저장되는지 확인하는 테스트 구현
